### PR TITLE
Support to defer to send a response in services

### PR DESCRIPTION
**Pull request by ivanpauno**
Fixes https://github.com/ros2/rclcpp/issues/1707, https://github.com/ros2/rclcpp/issues/491.

TODO: add tests

### CODE REVIEW

File: rclcpp/include/rclcpp/any_service_callback.hpp [hunk 1]
```diff
@@ -29,93 +31,116 @@
...
+      std::shared_ptr<typename ServiceT::Response>
+    )>;
+  using SharedPtrDeferResponseCallback = std::function<
+    void (
+      std::shared_ptr<rmw_request_id_t>,
+      std::shared_ptr<typename ServiceT::Request>
+    )>;
+
+  std::variant<
+    std::monostate,
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/any_service_callback.hpp:140]: Is the monostate needed if the callbacks can be default initialized?

**Review by wjwwood — APPROVED** [review]
lgtm, with one question.

File: rclcpp/include/rclcpp/any_service_callback.hpp (hunk 1 above)
Comment (ivanpauno) [reply rclcpp/include/rclcpp/any_service_callback.hpp:140]: It's not needed.
The approach I used is to use `std::monostate` as the default, and to check that the callbacks passed to `set()` as an argument are not null.
e.g. if `callback_` is in the `SharedPtrCallback` variant, I don't need to check if the callback is actually `nullptr` or not, as that was checked before.

I think that `std::monostate` just use an extra value of the variant discriminant, so implementation wise the difference doesn't matter much.

**Comment by ivanpauno** [comment]
> Is the TODO done? I guess not, only tests have been updated, but not new ones added.

Not yet, I will add some tests before running CI.

**Comment by ivanpauno** [comment]
All client/service tests were in `test_rclcpp`, so I have opened https://github.com/ros2/system_tests/pull/478.

I forgot that in most cases you actuall also want to receive a `std::shared_ptr<Service>` in the callback, for e.g. doing things like this:

https://github.com/ivanpauno/domain_bridge/blob/5377aaf95a1b812c8d1df52663a1e1ca5feb6ca3/include/domain_bridge/service_bridge_impl.inc#L181-L195

For being able to do this, we would need to add `std::enable_shared_from_this` as a base class of `Service`, @wjwwood do you think that would be okay?
You can also work around that problem by creating a simpler wrapper class, similar of what I did in the `test_rclcpp` PR, but having direct support for that in rclcpp is maybe not a bad idea.

**Comment by wjwwood** [comment]
Well, the user can always capture the service in a lambda or use `std::bind`. They also have to keep the service in scope anyway, so I don't think we need to do it for them.

That being said, we can have the enable shared from this, but maybe we should avoid it if we can.

**Comment by ivanpauno** [comment]
> Well, the user can always capture the service in a lambda or use std::bind. They also have to keep the service in scope anyway, so I don't think we need to do it for them.

The problem is that they can't because they have to pass the lambda to `create_service()`, which is the method returning the `service` object, so there's a chicken and egg problem.

**Comment by wjwwood** [comment]
> The problem is that they can't because they have to pass the lambda to create_service(), which is the method returning the service object, so there's a chicken and egg problem.

Ah, I see. Hmm, in that case I suppose it would be better to do as you say.

**Review by wjwwood — COMMENTED** [review]
lgtm, waiting to approve until you make the change you mentioned.

**Comment by ivanpauno** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14903/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9646/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12572/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15058/)

**Review by wjwwood — APPROVED** [review]

**Comment by ivanpauno** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14911/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9653/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12581/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15068/)

**Comment by ivanpauno** [comment]
Windows CI again:

* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15069/)

File: rclcpp/test/rclcpp/test_any_service_callback.cpp [hunk 2]
```diff
@@ -73,6 +74,36 @@ TEST_F(TestAnyServiceCallback, set_and_dispatch_header) {
 
   any_service_callback_.set(callback_with_header);
   EXPECT_NO_THROW(
-    any_service_callback_.dispatch(request_header_, request_, response_));
+    EXPECT_NE(nullptr, any_service_callback_.dispatch(nullptr, request_header_, request_)));
+  EXPECT_EQ(callback_with_header_calls, 1);
+}
+
+TEST_F(TestAnyServiceCallback, set_and_dispatch_defered) {
```
Comment (aboseley) [inline rclcpp/test/rclcpp/test_any_service_callback.cpp:81]: It would be great if an example was added to ros/examples for this new feature

**Comment by ros-discourse** [comment]
This issue has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/deferrable-canceleable-lifecycle-transitions/32318/1

File: rclcpp/test/rclcpp/test_any_service_callback.cpp (hunk 2 above)
Comment (tgroechel) [reply rclcpp/test/rclcpp/test_any_service_callback.cpp:81]: For anyone coming back to this later, I threw together an example of async services here: https://github.com/tgroechel/lifecycle_prac/blob/main/src/async_srv.cpp#L70-L108

Haven't compiled it yet but should give the idea (I patched this together while doing deferrable lifecycle transitions)

Comment (tgroechel) [reply rclcpp/test/rclcpp/test_any_service_callback.cpp:81]: And here is an example (also not compiled) leveraging the `SharedFutureResponse` in order to call a service within a service: https://github.com/tgroechel/lifecycle_prac/blob/main/src/async_srv.cpp#L10-L69C1

**Pull request merged**